src/System.Private.CoreLib/shared/System/Threading/WaitHandle.cs

```diff
@@ -0,0 +1,384 @@
+                }
+                else
+                {
+                    waitResult = WaitOneCore(waitHandle.DangerousGetHandle(), millisecondsTimeout);
+                }
+
+                if (waitResult == WaitAbandoned)
+                {
+                    throw new AbandonedMutexException();
+                }
+
+                return waitResult != WaitTimeout;
+            }
+            finally
+            {
+                if (success)
+                    waitHandle.DangerousRelease();
+            }
+        }
+
+        /// <summary>
+        /// Obtains all of the corresponding safe wait handles and adds a ref to each. Since the <see cref="SafeWaitHandle"/>
+        /// property is publically modifiable, this makes sure that we add and release refs one the same set of safe wait
+        /// handles to keep them alive during a multi-wait operation.
+        /// </summary>
+        private static void ObtainSafeWaitHandles(
+            ReadOnlySpan<WaitHandle> waitHandles,
+            out SafeWaitHandle[] safeWaitHandles,
+            out IntPtr[] unsafeWaitHandles)
+        {
+            Debug.Assert(waitHandles != null);
+            Debug.Assert(waitHandles.Length > 0);
+            Debug.Assert(waitHandles.Length <= MaxWaitHandles);
+
+            safeWaitHandles = new SafeWaitHandle[waitHandles.Length];
+            unsafeWaitHandles = new IntPtr[waitHandles.Length];
+            bool success = false;
+            try
+            {
+                for (int i = 0; i < waitHandles.Length; ++i)
+                {
+                    WaitHandle waitHandle = waitHandles[i];
+                    if (waitHandle == null)
+                    {
+                        throw new ArgumentNullException("waitHandles[" + i + ']', SR.ArgumentNull_ArrayElement);
+                    }
+
+                    SafeWaitHandle safeWaitHandle = waitHandle._waitHandle;
+                    if (safeWaitHandle == null)
+                    {
+                        // Throw ObjectDisposedException for backward compatibility even though it is not be representative of the issue
+                        throw new ObjectDisposedException(null, SR.ObjectDisposed_Generic);
+                    }
+
+                    safeWaitHandle.DangerousAddRef(ref success);
+                    safeWaitHandles[i] = safeWaitHandle;
+                    unsafeWaitHandles[i] = safeWaitHandle.DangerousGetHandle();
+                }
+                success = true;
+            }
+            finally
+            {
+                if (!success)
+                {
+                    for (int i = 0; i < waitHandles.Length; ++i)
+                    {
+                        SafeWaitHandle safeWaitHandle = safeWaitHandles[i];
+                        if (safeWaitHandle == null)
+                        {
+                            break;
+                        }
+                        safeWaitHandle.DangerousRelease();
```

Comment:
Should this release only handles for which DangerousAddRef succeeded?

Reply:
Kinda. The thing is that `DangerousAddRef` always throws on error. I believe the whole thing with `ref success` is about preventing a race condition on thread abort which cannot really be triggered by the user on .NET Core anymore. I can probably rewrite it to respect the last `success` value (and keep a reference to last `SafeWaitHandle` before `safeWaitHandle.DangerousAddRef`) and assume it was `true` if `safeWaitHandles[i]` is non-null.